hawk/app/models/report.rb

```diff
@@ -1,69 +1,210 @@
+          v = peinput_version path
+          if v && v != pcmk_version
+            version = _("PE Input created by different Pacemaker version (%{version})" % { :version => v })
+          elsif v != pcmk_version
+            version = _("Pacemaker version not present in PE Input")
+          else
+            version = nil
+          end
+          peinputs.push(timestamp: File.mtime(path).iso8601,
+                        basename: File.basename(path, ".bz2"),
+                        filename: File.basename(path),
+                        path: path.sub("#{hb_report.path}/", ''),
+                        node: path.split(File::SEPARATOR)[-3],
+                        version: version)
+        end
+        # sort is going to be off for identical mtimes (stripped back to the second),
+        # so need secondary sort by filename
+      end
+
+      # add errors to output
+      errors = hb_report.err_filtered
+      errors.each do |err|
+        peinputs.push error: err
+      end
+    end
+  end
 
+  def peinput_version(path)
+    nvpair = `CIB_file=#{path} cibadmin -Q --xpath "/cib/configuration//crm_config//nvpair[@name='dc-version']" 2>/dev/null`
+    m = nvpair.match(/value="([^"]+)"/)
+    return nil unless m
+    m[1]
   end
 
-  def report_path
-    self.class.report_path
+  def transition_cmd(hb_report, path, cmd)
+    source = archive
+    source = hb_report.path if File.directory?(hb_report.path)
+    Util.capture3("crm", "history", stdin_data: "source #{source}\n#{cmd}\n")
   end
 
-  class << self
-    def parse(file)
-      hash = Digest::SHA1.hexdigest(
-        file.basename(".tar.bz2").to_s
-      )
+  def info(hb_report, path)
+    out, err, status = transition_cmd hb_report, path, "transition #{path} nograph"
+    info = out + err
+    info.strip!
+    info = _("No details available") if info.empty?
+    info.insert(0, _("Error:") + "\n") unless status.exitstatus == 0
+    info
+  end
 
-      match = file.basename.to_s.match(
-        /(\d{4}-\d{2}-\d{2}_\d{2}:\d{2})/
-      )
+  def tags(hb_report, path)
+    out, err, status = transition_cmd hb_report, path, "transition tags #{path}"
+    out.split
+  end
 
-      order = [
-        match[0],
-        match[1]
-      ].join("-")
+  def logs(hb_report, path)
+    out, err, status = transition_cmd hb_report, path, "transition log #{path}"
+    info = out + err
+    info.strip!
+    info = _("No details available") if info.empty?
+    info.insert(0, _("Error:") + "\n") unless status.exitstatus == 0
+    info
+  end
 
-      Report.new(
-        id: hash,
-        order: order,
+  # Apparently we can't rely on the dot file existing in the hb_report, so we
+  # just use ptest to generate it.  Note that this will fail if hacluster doesn't
+  # have read access to the pengine files (although, this should be OK, because
+  # they're created by hacluster by default).
+  # Returns [success, data|error]
+  def graph(hb_report, path, format = :svg)
```

Comment:
Metrics/AbcSize: Assignment Branch Condition size for graph is too high. [30.1/15]